Add support for QEMU virt ARMv8-A

jenswi-linaro opened the pull request:
Tested with among other flags for QEMU:
```
qemu-system-aarch64 -machine virt,secure=on -cpu cortex-a57 \
        -smp 2 -m 1024 -semihosting-config enable,target=native
```

danh-arm commented on plat/qemu/qemu_bl2_setup.c:
```diff
@@ -0,0 +1,307 @@
+void bl2_plat_get_bl30_meminfo(meminfo_t *bl30_meminfo)
+{
+	*bl30_meminfo = bl2_tzram_layout;
+}
```
References to BL30 (or the new new name, SCP_BL2) can probably be removed (not just here but in other places).

danh-arm commented on plat/qemu/qemu_bl31_setup.c:
```diff
@@ -0,0 +1,167 @@
+
+uint64_t plat_get_syscnt_freq(void)
+{
+	return (1000 * 1000 * 1000) / 16;
```
minor: magic numbers discouraged

danh-arm commented on plat/qemu/qemu_io_storage.c:
```diff
@@ -0,0 +1,395 @@
+		}
+	}
+	return result;
+}
```
Does semihosting really work on QEMU? If not tested, should be removed.

jenswi-linaro replied:
Yes, that's what I'm using when testing.

danh-arm commented on include/libfdt/fdt.h:
```diff
@@ -0,0 +1,111 @@
+#ifndef _FDT_H
```
It's better (for legal reasons) if this 3rd party IP is imported into the codebase by us (ARM). We already have approval for importing libfdt so I will arrange this. Then you will be able to remove this commit from the PR.

jenswi-linaro replied:
OK, I'll remove it when rebasing.

danh-arm commented on plat/qemu/qemu_bl2_setup.c:
```diff
@@ -32,6 +32,9 @@
 #include <string.h>
+#include <libfdt.h>
+#include <debug.h>
+#include <psci.h>
```
minor: prefer #includes to be alphabetically listed.

danh-arm commented on plat/qemu/qemu_bl2_setup.c:
```diff
@@ -175,9 +178,120 @@ static void security_setup(void)
+	if (ret < 0)
+		ERROR("Failed to pack Device Tree at %p: error %d\n", fdt, ret);
+}
+
```
This PSCI DT node injection code is potentially useful for other platforms. It might be worth pulling these utility functions out into a separate file (within the platform port) to faciliate this in future.

danh-arm commented:
A small file in plat/doc would be useful, explaining how to build, use and get more information.

jenswi-linaro commented on plat/qemu/qemu_pm.c:
```diff
@@ -183,7 +192,14 @@ void qemu_pwr_domain_suspend(const psci_power_state_t *target_state)
+	gicv2_pcpu_distif_init();
+
+	/* Enable the gic cpu interface */
+	gicv2_cpuif_enable();
```
Yes, I hope it can be done in a later P-R.

danh-arm commented:
I've just pushed #639 to import libfdt. The isn't exactly the same as your patch but hopefully should work for you. Let me know if you see any issues with it.

danh-arm commented:
Hi @jenswi-linaro . I'm going to merge #639 this week unless you have any objections. Regards, Dan.

jenswi-linaro commented:
Thanks @danh-arm, #639 works fine with my patches.

jenswi-linaro commented:
Review comments addressed and rebased on integration branch.
Squashed the the commits "Add support for QEMU virt ARMv8-A target", "qemu: add smp support" and "qemu: edit device tree" into "Add support for QEMU virt ARMv8-A target".

danh-arm commented:
jenkins:test this please
